### PR TITLE
💉  Request and share immunisation data

### LINKED ISSUE #178: Document stays shared after user removes it from their records
If I remove a document from my records, shares of that document should be revoked

### PR DESCRIPTION
**Pull request by mermop**
Add ImmunisationRecords:

- Add a new type of Document - ImmunisationRecord
- Add shared tests for Document
- Add documentation for creating more new types of Document

Genericise a number of things that were previously just applicable to BirthRecord:

- Remove user `birth_records#show` in favour of `documents#show`
- Remove user `birth_records#remove` in favour of `documents#remove`
- Remove user `birth_records#index` in favour of `dashboard#index`

Misc fixes:

- Revoke shares when a document is removed - fixes #178 

Stories:

- [Organisation requesting immunisation data](https://trello.com/c/Prg1tZFg/77-organisation-requests-immunisation-data-from-parent)
- [Parent sharing immunisation record](https://trello.com/c/vmKyxwTE/79-parent-shares-immunisation-data)

Does not include:

- [Parent adds immunisation record](https://trello.com/c/OPO1Twou/78-parent-adds-immunisation-data)

### CODE REVIEW

File: app/views/user/shares/show.html.erb [hunk 1]
```diff
@@ -14,5 +12,11 @@
...
   <%= @share.recipient.name %>
 </p>
-
-<%= link_to 'Back', user_shares_path %>
+<div class="card">
+  <div class="card-header">
+    <h5>Document details</h3>
+    </div>
+  <div class="card-body">
+    <%= render @share.document %>
```
[the remark this answers is not in the document]
Comment (mermop) [reply app/views/user/shares/show.html.erb:20]: What

[the remark this answers is not in the document]
Comment (mermop) [reply app/views/user/shares/show.html.erb:20]: Brakeman false positive https://github.com/presidentbeef/brakeman/pull/529

File: config/environments/test.rb [hunk 2]
```diff
@@ -46,7 +46,4 @@
   # config.action_view.raise_on_missing_translations = true
 
   Faker::Config.locale = 'en-NZ'
-  # Set Faker seed so that it creates names in deterministic order
-  # since we are using Percy for visual diffs
-  Faker::Config.random = Random.new(42)
```
Comment (mermop) [inline config/environments/test.rb:51]: I'm setting this freshly in relevant tests

File: config/routes.rb [hunk 3]
```diff
@@ -23,6 +23,9 @@
 
   resources :user, only: [:index]
   namespace :user do
+    resources :dashboard, only: %i[index]
+    get 'documents/:type/:id', to: 'documents#show', as: :document
+    post 'documents/:type/:id/remove', to: 'documents#remove', as: :document_remove
```
Comment (mermop) [inline config/routes.rb:12]: There's gotta be a tidier way to do this 😬

File: spec/factories/immunisation_record_data.json [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+    age_due: 'P15M',
+    vaccine: 'VV'
+  },
+  {
+    age_due: 'P4Y',
+    vaccine: 'DTaP-IPV'
+  },
+  {
+    age_due: 'P4Y',
+    vaccine: 'MMR'
```
Comment (mermop) [inline spec/factories/immunisation_record_data.json]: Initial thoughts about data format - doesn't need to be checked in this branch

File: config/routes.rb (hunk 3 above)
Comment (mermop) [reply config/routes.rb:12]: @Br3nda is making some other changes to the routes in #172 so I'm gonna say that we should have a wholesale rethink of them once both of these are merged

File: Gemfile [hunk 5]
```diff
@@ -61,9 +61,13 @@ gem 'discard', '~> 1.0'
 gem 'state_machines'
 gem 'state_machines-activerecord'
 
+# HTTP connections
 gem 'faraday'
 gem 'faraday_middleware'
 
+# Validating NHI number
+gem 'nhi-validator', '~> 0.1.0'
```
Comment (Br3nda) [inline Gemfile:66]: oooooooo

File: app/models/concerns/document.rb [hunk 6]
```diff
@@ -4,11 +4,25 @@
...
+    has_many :user_documents, dependent: :nullify, as: :document
     has_many :users, -> { distinct }, through: :user_documents
-    has_many :shares, -> { merge(Share.kept) }, dependent: :nullify, inverse_of: :document, foreign_key: 'document_id'
+    has_many :shares, -> { merge(Share.kept) }, as: :document, dependent: :nullify,
+                                                inverse_of: :document, foreign_key: 'document_id'
+
+    delegate :share_audit_comment, :remove_audit_comment, :add_audit_comment, :view_audit_comment, to: :class
+
+    def immunisation_record?
+      document_type == IMMUNISATION_RECORD
```
Comment (Br3nda) [inline app/models/concerns/document.rb:20]: this line isn't covered in the tests.  not used?

File: app/services/audited_operations_service.rb [hunk 7]
```diff
@@ -62,23 +60,23 @@ def self.revoke_share(user:, share:)
...
-  def self.access_shared_birth_record(logged_identity:, share:)
+  def self.access_shared_document(user:, share:)
     raise ArgumentError, 'share cannot be nil' if share.nil?
-    raise ArgumentError, 'logged_identity cannot be nil' if logged_identity.nil?
+    raise ArgumentError, 'user cannot be nil' if user.nil?
     raise ShareRevokedError if share.revoked?
 
     # update last_accessed_at, which will generate an audit record
-    Audited.audit_class.as_user(logged_identity) do
+    Audit.as_user(user) do
```
Comment (Br3nda) [inline app/services/audited_operations_service.rb:69]: nicer!

File: app/validators/national_health_index_validator.rb [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+
+require 'nhi_validator'
+
+class NationalHealthIndexValidator < ActiveModel::Validator
+  def validate(record)
+    record.errors[:nhi] << (options[:message] || 'is not a valid NHI') unless NHIValidator.valid?(record.nhi)
+  end
+
+  def validate_each(record, attribute, value)
+    record.errors[attribute] << (options[:message] || 'is not a valid NHI') unless NHIValidator.valid?(value)
```
Comment (Br3nda) [inline app/validators/national_health_index_validator.rb]: no tests for this line either

**Comment by mermop** [comment]
I don't see the microbe emoji on my machine!

File: app/models/concerns/document.rb (hunk 6 above)
Comment (mermop) [reply app/models/concerns/document.rb:20]: hmm, should be used in `shares/index`

File: app/validators/national_health_index_validator.rb (hunk 8 above)
Comment (mermop) [reply app/validators/national_health_index_validator.rb]: Right - we don't actually need the `validate_each` section, I'll remove it

File: app/controllers/user/shares_controller.rb [hunk 9]
```diff
@@ -33,27 +33,24 @@ def create
...
-    birth_record = current_user.birth_records.find_by(id: share_params['document_id'])
-    authorize birth_record, :share?
-    recipient = Organisation.find_by(id: share_params['recipient_id'])
-    birth_record.share_with(
-      recipient: recipient,
-      user: current_user
-    )
+    document = current_user.documents(type: share_params[:document_type])
+                           .find_by(id: share_params[:document_id])
+    authorize document, :share?
```
Comment (Br3nda) [inline app/controllers/user/shares_controller.rb]: i wonder if this should be `authorize @share, :create?` instead.
the pundit would check that you're sharing a document that is yours.

Comment (Br3nda) [reply app/controllers/user/shares_controller.rb]: it's nice when controllers are straight forward
```
@share = Share.new(share_params)
authorize @show
@share.save!
```

Comment (Br3nda) [reply app/controllers/user/shares_controller.rb]: (but then the auditing... i've usualy put that on the model `after_save`)

**Review by Br3nda — APPROVED** [review]

File: app/controllers/user/shares_controller.rb (hunk 9 above)
Comment (mermop) [reply app/controllers/user/shares_controller.rb]: have updated that now

File: app/controllers/user/shares_controller.rb [hunk 10]
```diff
@@ -29,52 +29,46 @@ def new
...
+
   # Require the params which will allow a valid model to be created
   #
   # The return value isn't useful, but this will raise an error if parameters
   # are missing
   def require_valid_params
     params
       .require(:share)
-      .require([:document_id, :recipient_id])
+      .require(%i[document_id document_type recipient_id])
```
Comment (Br3nda) [inline app/controllers/user/shares_controller.rb:63]: why do we need this method? Won't the model validators and pundit permissions do this for us?

**Pull request merged**